src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs

```diff
@@ -156,87 +56,69 @@ public override int Order
             {
                 throw new ArgumentNullException(nameof(output));
             }
+            
+            IHtmlContent content = null;
 
-            IHtmlContent result = null;
             if (Enabled)
             {
                 var key = GenerateKey(context);
-                if (!MemoryCache.TryGetValue(key, out result))
+                MemoryCacheEntryOptions options;
+    
+                while(content == null)
                 {
-                    // Create an entry link scope and flow it so that any tokens related to the cache entries
-                    // created within this scope get copied to this scope.
-                    using (var link = MemoryCache.CreateLinkingScope())
+                    Task<IHtmlContent> result = null;
+                    
+                    if (!MemoryCache.TryGetValue(key, out result))
                     {
-                        var content = await output.GetChildContentAsync();
-
-                        var stringBuilder = new StringBuilder();
-                        using (var writer = new StringWriter(stringBuilder))
+                        // Create an entry link scope and flow it so that any tokens related to the cache entries
+                        // created within this scope get copied to this scope.
+                        options = GetMemoryCacheEntryOptions();
+
+                        var tcs = new TaskCompletionSource<IHtmlContent>();
+                         
+                        MemoryCache.Set(key, tcs.Task, new MemoryCacheEntryOptions());
+                        
+                        try 
                         {
-                            content.WriteTo(writer, HtmlEncoder);
+                            using (var link = MemoryCache.CreateLinkingScope())
+                            {
+                                result = ProcessContentAsync(output);
+                                content = await result;
+                                options.AddEntryLink(link);
+                            }
+                            
+                            // The entry is set instead of assigning a value to the 
+                            // task so that the expiration options are are not impacted 
+                            // by the time it took to compute it.
+                            
+                            MemoryCache.Set(key, result, options);
+                            tcs.SetResult(null);
+                        }
+                        catch (Exception ex)
+                        {
+                            tcs.TrySetException(ex); 
```

Comment:
Should the other awaiters fail, or should they be given a SetResult(null) so they can take a crack at rendering?